### PR TITLE
Multistore - International > Localization > Geolocation - Add checkboxes

### LINKED ISSUE #19323: Multistore - International > Localization > Geolocation - Add specific settings dropdown
As a merchant, I want to be able to see if a parameter has a different value for another store when I am in "all shops" context or in a group context, so I can know to which stores the change I am making will be applied

[Specs specific settings dropdown](https://github.com/PrestaShop/prestashop-specs/blob/9783996765a46d3d7197e4f7cd7098ffb7e37ff2/content/1.7/back-office/multistoregeneralspecs.md#specific-settings-drop-down-only-from-178)

### LINKED ISSUE #19371: Multistore - International > Localization > Geolocation - Add checkboxes
As a merchant, I want to be able to select one or many fields / toggles when I am in a specific shop context or in a group context, so I can apply a different configuration for the selected store or group and see which parameters have different value(s) for the selected store or group.

[Specs checkboxes](https://github.com/PrestaShop/prestashop-specs/blob/master/back-office/multistoregeneralspecs.md#checkboxes)

### PR DESCRIPTION
**Pull request by ghost**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | As a merchant, I want to be able to select one or many fields / toggles when I am in a specific shop context or in a group context, so I can apply a different configuration for the selected store or group and see which parameters have different value(s) for the selected store or group.
| Type?             | improvement
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     | no
| Fixed ticket?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/19371 Fixes https://github.com/PrestaShop/PrestaShop/issues/19323
| How to test?      | See issue [#27540](https://github.com/PrestaShop/PrestaShop/issues/27540) create a fake GeoLite2-City.mmdb in app ressource geolite
| Possible impacts? | Please indicate what parts of the software we need to check to make sure everything is alright.

:hammer: **Changed configuration forms and unit tests**

- [x] GeolocationByIpAddressConfiguration Form
- [x] Add unit test for GeolocationByIpAddressConfiguration
- [x] GeolocationIpAddressWhitelistConfiguration Form
- [x] Add unit test GeolocationByIpAddressConfiguration
- [x] GeolocationOptionsConfiguration Form
- [x] Add unit test for GeolocationOptionsConfiguration

:information_source:  **Informations BC Break**

- Constructor `GeolocationByIpAddressConfiguration` changes
- Constructor `GeolocationIpAddressWhitelistConfiguration` changes
- Constructor `GeolocationOptionsConfiguration` changes

:camera: **Screenshot**

Comming soon

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27533)

### CODE REVIEW

File: src/Adapter/Geolocation/GeolocationByIpAddressConfiguration.php [hunk 1]
```diff
@@ -26,55 +26,47 @@
...
-
-    /**
-     * @param Configuration $configuration
-     */
-    public function __construct(
-        Configuration $configuration
-    ) {
-        $this->configuration = $configuration;
-    }
+    private const CONFIGURATION_FIELDS = ['geolocation_enabled'];
```
Comment (saulaski) [inline src/Adapter/Geolocation/GeolocationByIpAddressConfiguration.php]: What about constants to reuse in the entire class ? easier to maintain :-)
```suggestion
    private const CONFIGURATION_KEY = 'PS_GEOLOCATION_ENABLED';
    private const CONFIGURATION_FIELD = 'geolocation_enabled';
    private const CONFIGURATION_FIELDS = [self::CONFIGURATION_FIELD];
```
```
    public function getConfiguration()
    {
        return [
            self::CONFIGURATION_FIELD => (bool) $this->configuration->get(self::CONFIGURATION_KEY, false, $this->getShopConstraint()),
        ];
    }

    public function updateConfiguration(array $configuration)
    {
        if ($this->validateConfiguration($configuration)) {
            $this->updateConfigurationValue(self::CONFIGURATION_KEY, self::CONFIGURATION_FIELD, $configuration, $this->getShopConstraint());
        }

        return [];
    }

    protected function buildResolver(): OptionsResolver
    {
        $resolver = (new OptionsResolver())
            ->setDefined(self::CONFIGURATION_FIELDS)
            ->setAllowedTypes(self::CONFIGURATION_FIELD, 'bool');

        return $resolver;
    }
```

**Review by saulaski — DISMISSED** [review]

File: src/Adapter/Geolocation/GeolocationByIpAddressConfiguration.php (hunk 1 above)
Comment (ghost) [reply src/Adapter/Geolocation/GeolocationByIpAddressConfiguration.php]: Thanks, I'll look into it when the PR is no longer blocked by : https://github.com/PrestaShop/PrestaShop/issues/27540

Regards

Comment (saulaski) [reply src/Adapter/Geolocation/GeolocationByIpAddressConfiguration.php]: I've commented only 1 class, but my suggestion applies to all classes... not an absolute requirement though :)

Regards

Comment (ghost) [reply src/Adapter/Geolocation/GeolocationByIpAddressConfiguration.php]: I will also look at my already merged PRs

**Review by saulaski — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by matthieu-rolland — DISMISSED** [review]

**Review by HanaRebaiQA — CHANGES_REQUESTED** [review]
Hello @okom3pom 

I found these issues while testing this PR.

1. Enable multistore > Add new shops
2. BO > IMPROVE > International > Geolocation
3. Select a checkboxes for a shop and click on Save (For example for Options > Choose `Geolocation behavior for other countries` > Save)
4. Select All shops 
5. Click on the specfic settings dropdown 
6. See errors

**Issue 1** : The list of the dorpdown is not diplayed and an error in the console
image
image


**Issue 2** : A bad display of the option element
image

These issues appear for the first selected checkbox. 
If we select more than one checkbox for a shop, then only the firt one will have a problem. 
If we select only one checkbox for a shop, this one will have a problem.

Another example : 
image


https://user-images.githubusercontent.com/92912932/155501014-78c2d5bf-4de4-42ee-83ed-78dca214b6d0.mp4

Could you please check it ?

Thanks!

**Comment by ghost** [comment]
Have you run `make` and clear cache ?

**Comment by ghost** [comment]
I found why : _Since December 30, 2019, you need to register for a [MaxMind](https://dev.maxmind.com/geoip/geoip2/geolite2)[ account to get a license key to be able to download the geolocation data. Once downloaded, extract the data using Winrar or Gzip into the /app/Resources/geoip/ directory.]()_ 

If you have geolite2 no issue i have to check

**Review by HanaRebaiQA — COMMENTED** [review]
@okom3pom Yes! i did a make install during the PR installation.
1. Clear cache browser
2. Clear cache PS from BO > CONFIGURE > Advanced Parameters > Performance > Choose `Force Compilation` > Save Then Click On `Clear cache`
3. Trying with private navigation 

https://user-images.githubusercontent.com/92912932/155506237-84d4d740-eaf6-497d-908b-10d72371084c.mp4

Thanks!

**Comment by ghost** [comment]
> I found why : _Since December 30, 2019, you need to register for a [MaxMind](https://dev.maxmind.com/geoip/geoip2/geolite2) account to get a license key to be able to download the geolocation data. Once downloaded, extract the data using Winrar or Gzip into the /app/Resources/geoip/ directory._
> 
> If you have geolite2 no issue i have to check

**Review by Progi1984 — APPROVED** [review]

**Comment by ghost** [comment]
I update how to test

**Review by matthieu-rolland — APPROVED** [review]

**Review by HanaRebaiQA — CHANGES_REQUESTED** [review]
Hello @okom3pom 

I checked again this PR. The Localization page is completely empty with multistore enabled and disabled **without using the geolite**.

https://user-images.githubusercontent.com/92912932/156589307-23082976-c54b-4cc2-9a1e-da22bb6f6dfd.mp4

Without PR : It was OK (without geolite)

https://user-images.githubusercontent.com/92912932/156590733-2d0f1a3f-2f8f-47b0-a3b8-f041e723251e.mp4

And when i used `GeoLite2-Country.mmdb `and `GeoLite2-City.mmdb` files under app/Resources/geoip, i have these issues : 
-Enable multistore
-Click on view shop for the new shop created under the default group
image
-Click on view my shop of the new shop created under a new shop group 
image

https://user-images.githubusercontent.com/92912932/156602321-0da2848c-10bd-40b2-829a-0667b553a013.mp4

This problem did not exists without PR.

https://user-images.githubusercontent.com/92912932/156602774-63c32e24-cdbb-4004-94df-60b6d27c5369.mp4

Could you please check and feedback ?

Thanks!

**Pull request closed without merging**